tus/src/main/java/ch/cyberduck/core/tus/TusUploadFeature.java

```diff
@@ -0,0 +1,250 @@
+import ch.cyberduck.core.preferences.PreferencesFactory;
+import ch.cyberduck.core.shared.DefaultUrlProvider;
+import ch.cyberduck.core.threading.BackgroundExceptionCallable;
+import ch.cyberduck.core.threading.DefaultRetryCallable;
+import ch.cyberduck.core.transfer.TransferStatus;
+
+import org.apache.commons.codec.binary.Base64;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.concurrent.ConcurrentUtils;
+import org.apache.http.Header;
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpResponse;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.HttpClient;
+import org.apache.http.client.HttpResponseException;
+import org.apache.http.client.ResponseHandler;
+import org.apache.http.client.methods.HttpHead;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Future;
+
+import static ch.cyberduck.core.tus.TusCapabilities.*;
+
+public class TusUploadFeature extends HttpUploadFeature<Void, MessageDigest> {
+    private static final Logger log = LogManager.getLogger(TusUploadFeature.class);
+
+    public static final String UPLOAD_URL = "uploadUrl";
+
+    private final Host host;
+    private final HttpClient client;
+    private final Preferences preferences = PreferencesFactory.get();
+
+    private Write<Void> writer;
+    private final TusCapabilities capabilities;
+
+    public TusUploadFeature(final Host host, final HttpClient client, final Write<Void> writer, final TusCapabilities capabilities) {
+        super(writer);
+        this.host = host;
+        this.client = client;
+        this.writer = writer;
+        this.capabilities = capabilities;
+    }
+
+    @Override
+    public Void upload(final Path file, final Local local, final BandwidthThrottle throttle, final StreamListener listener,
+                       final TransferStatus status, final ConnectionCallback callback) throws BackgroundException {
+        // In order to achieve parallel upload the Concatenation extension MAY be used.
+        try {
+            final List<Future<Void>> chunks = new ArrayList<>();
+            long offset = status.getOffset();
+            long remaining = status.getLength();
+            final String uploadUrl;
+            if(status.isAppend()) {
+                uploadUrl = preferences.getProperty(toUploadUrlPropertyKey(host, file));
+                if(log.isDebugEnabled()) {
+                    log.debug(String.format("Resume upload to %s for %s from offset %d", uploadUrl, file, status.getOffset()));
+                }
+            }
+            else {
+                if(!capabilities.extensions.contains(Extension.creation)) {
```

Comment:
Move check to `Session#connect` to determine which `Upload` feature to take - based on version and creation extension. If the check is here only the client is unable to upload anything to a server with TUS support but no creation extension impl.